### PR TITLE
RDBC-772 Cluster wide transactions

### PR DESCRIPTION
**Pull request by maciejaszyk**
Issue: https://issues.hibernatingrhinos.com/issue/RDBC-772/Support-for-cluster-wide-transactions-for-Go-client

Support for:
- cluster wide session
- creating compare exchanges within session
- deleting compare exchanges within session

### CODE REVIEW

File: session_options.go [hunk 1]
```diff
@@ -2,6 +2,16 @@ package ravendb
 
 // SessionOptions describes session options
 type SessionOptions struct {
-	Database        string
-	RequestExecutor *RequestExecutor
+	Database                                            string
+	RequestExecutor                                     *RequestExecutor
+	TransactionMode                                     int
```
Comment (maciejaszyk) [inline session_options.go:7]: Another

File: batch_command.go [hunk 2]
```diff
@@ -23,25 +23,42 @@ type BatchCommand struct {
 	commands          []ICommandData
 	attachmentStreams []io.Reader
 	options           *BatchOptions
+	Result            *JSONArrayResult
 
-	Result *JSONArrayResult
+	transactionMode             int
+	disableAtomicDocumentWrites *bool
+	raftUniqueRequestId         string
```
Comment (maciejaszyk) [inline batch_command.go:30]: In go `BatchCommand` and `ClusterWideCommand` are implemented via `BatchCommand` and if requires we build request based on this properties.

File: batch_operation.go [hunk 3]
```diff
@@ -31,10 +31,94 @@ func (b *BatchOperation) createRequest() (*BatchCommand, error) {
 
 	b.entities = result.entities
 
-	return newBatchCommand(b.session.GetConventions(), result.sessionCommands, result.options)
+	return newBatchCommand(b.session.GetConventions(), result.sessionCommands, result.options, b.session.transactionMode, b.session.disableAtomicDocumentWritesInClusterWideTransaction) //todo disableAtomic!!!
```
[the remark this answers is not in the document]
Comment (maciejaszyk) [reply batch_operation.go:34]: done

File: in_memory_document_session_operations.go [hunk 4]
```diff
@@ -84,9 +84,16 @@ type InMemoryDocumentSessionOperations struct {
 	// so we can upcast/downcast between them
 	// In Go we need a backlink to reach DocumentSession
 	session *DocumentSession
+
+	//transaction configuration
+	transactionMode                                     int
+	disableAtomicDocumentWritesInClusterWideTransaction *bool
+	clusterTransactions                                 *ClusterTransactionOperations
+
```
[the remark this answers is not in the document]
Comment (maciejaszyk) [reply in_memory_document_session_operations.go:92]: Matches conventions from JVM client.

File: compare_exchange_session_value.go [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+		if hasChanged == false {
+			return nil, nil
+		}
+
+		if entityToInsert == nil {
+			entityToInsert = convertEntityToJSONRaw(entity, nil, false)
+			entityToInsert = cesv.convertEntity(metadata, entityToInsert)
+		}
+
+		return newPutCompareExchangeCommandData(cesv.key, entityToInsert, cesv.index), nil
```
[the remark this answers is not in the document]
Comment (maciejaszyk) [reply compare_exchange_session_value.go:216]: done

File: document_session_cluster_transaction.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
+package ravendb
+
+import (
+	"reflect"
+)
+
+type ClusterTransactionOperations struct {
+	session                             *InMemoryDocumentSessionOperations
+	state                               map[string]*CompareExchangeSessionValue
+	missingDocumentsTooAtomicGuardIndex map[string]string
```
Comment (ayende) [inline document_session_cluster_transaction.go:10]: Isn't it? 

```suggestion
	missingDocumentsToAtomicGuardIndex map[string]string
```

File: document_session_cluster_transaction.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+	value, exists := cto.missingDocumentsTooAtomicGuardIndex[docId]
+	if exists == false {
+		return nil, false
+	}
+
+	return &value, true
+}
+
+func (cto *ClusterTransactionOperations) IsTracked(key string) bool {
+	_, exists := cto.tryGetCompareExchangeValueFromSession(key)
```
Comment (ayende) [inline document_session_cluster_transaction.go:35]: Shouldn't we already check `missingDocumentsTooAtomicGuardIndex` ?

File: document_session_cluster_transaction.go [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+func (cto *ClusterTransactionOperations) getCompareExchangeValueFromSessionInternal(clazz reflect.Type, key string) (compareExchangeValue *CompareExchangeValue, notTracked bool) {
+	result, exist := cto.tryGetCompareExchangeValueFromSession(key)
+
+	if exist == false {
+		return nil, true
+	}
+
+	//we've already deserialized, maybe except situation when user wants get deriative type?
```
[the remark this answers is not in the document]
Comment (ayende) [reply document_session_cluster_transaction.go:268]: What happens if a user calls us twice with _different_ `clazz` values,  I guess?

File: document_session_cluster_transaction.go [hunk 9]
```diff
@@ -0,0 +1,328 @@
+package ravendb
+
+import (
+	"reflect"
+)
+
+type ClusterTransactionOperations struct {
+	session                             *InMemoryDocumentSessionOperations
+	state                               map[string]*CompareExchangeSessionValue
```
Comment (ayende) [inline document_session_cluster_transaction.go:9]: Aren't we missing `_compareExchangeIncludes` ?

File: readme.md [hunk 10]
```diff
@@ -950,3 +950,81 @@ case <-time.After(time.Second * 5):
...
+
+### Setup a session
+To set session transaction as cluster wide you've to set `TransactionMode` in `SessionOptions`
+as `TransactionMode_ClusterWide`
+
+```go
+ravendb.SessionOptions{
+                        [...]
+			TransactionMode: ravendb.TransactionMode_ClusterWide,
+}
```
Comment (ayende) [inline readme.md:964]: Shows how you pass that to `NewSessionWithOptions` here

File: readme.md [hunk 11]
```diff
@@ -950,3 +950,81 @@ case <-time.After(time.Second * 5):
...
+clusterTransaction, err := session.Advanced().ClusterTransaction()
+```
+
+Assert error to make sure your session configuration is correct.
+
+
+### Inserting new CompareExchangeValue
+
+```go
+var objectToInsert interface{}
```
Comment (ayende) [inline readme.md:982]: What did you actually insert here, a null, no?

File: readme.md [hunk 12]
```diff
@@ -950,3 +950,81 @@ case <-time.After(time.Second * 5):
...
+var id             string
+value, error := clusterTransaction.CreateCompareExchangeValue(key, objectToInsert)
+```
+
+### Getting existing CompareExchangeValue from server
+You can retrieve your value using various methods.
+
+#### - Get value by key
+```go
+var dataType reflect.Type // identifies your data-struct type
```
Comment (ayende) [inline readme.md:992]: That is null, no? What's is the actual value you passed here?

File: readme.md [hunk 13]
```diff
@@ -950,3 +950,81 @@ case <-time.After(time.Second * 5):
...
+var key string
+value, error := clusterTransaction.GetCompareExchangeValue(dataType, key)
+```
+
+#### - Get values by keys
+```go
+var dataType reflect.Type // identifies your data-struct type
+var keys []string
```
Comment (ayende) [inline readme.md:1000]: What keys are you loading? Assume that the user will copy & paste, we need actual examples here.

File: readme.md [hunk 14]
```diff
@@ -950,3 +950,81 @@ case <-time.After(time.Second * 5):
...
+```
+
+Retruns map where keys are identifiers.
+
+#### Get values whose IDs start with a string
+```go
+var dataType reflect.Type // identifies your data-struct type
+var startsWith string
+var start int
+var pageSize int
```
Comment (ayende) [inline readme.md:1011]: Same

File: readme.md (hunk 10 above)
Comment (maciejaszyk) [reply readme.md:964]: Done

File: document_session_cluster_transaction.go (hunk 7 above)
Comment (maciejaszyk) [reply document_session_cluster_transaction.go:35]: Well, I accidentally put it (I was using v5.2 JVM as reference implementing this but sometimes I checked C# client). I will skip `missingDocumentsToAtomicGuardIndex` for now and leave it for future development.

File: document_session_cluster_transaction.go (hunk 9 above)
Comment (maciejaszyk) [reply document_session_cluster_transaction.go]: We are, but we don't have any builder API for load so I decided not to write it from scratch in this PR, I've created a ticket for it: https://issues.hibernatingrhinos.com/issue/RDBC-786/Include-CompareExchangeValue-in-Load-method-in-go-client

**Pull request merged**